In [67]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, classification_report

In [35]:

df= pd.read_csv("/home/hgidea/Desktop/Coding/Python/internship/cognifz/Dataset .csv")
missing_values = df.isnull().sum()

In [36]:
print(missing_values)

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64


In [37]:
df.dropna(subset=['Cuisines'], inplace=True)
missing_values = df.isnull().sum()

In [38]:
print(missing_values)

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64


In [39]:
df = df[['Restaurant ID', 'Restaurant Name', 'Cuisines', 'Average Cost for two']]


In [40]:
df

Restaurant ID           Restaurant Name  \
0           6317637          Le Petit Souffle   
1           6304287          Izakaya Kikufuji   
2           6300002    Heat - Edsa Shangri-La   
3           6318506                      Ooma   
4           6314302               Sambo Kojin   
...             ...                       ...   
9546        5915730               Naml۱ Gurme   
9547        5908749              Ceviz A��ac۱   
9548        5915807                     Huqqa   
9549        5916112               A���k Kahve   
9550        5927402  Walter's Coffee Roastery   

                              Cuisines  Average Cost for two  
0           French, Japanese, Desserts                  1100  
1                             Japanese                  1200  
2     Seafood, Asian, Filipino, Indian                  4000  
3                      Japanese, Sushi                  1500  
4                     Japanese, Korean                  1500  
...                                ...                   ...  
9546                           Turkish                    80  
9547   World Cuisine, Patisserie, Cafe                   105  
9548            Italian, World Cuisine                   170  
9549                   Restaurant Cafe                   120  
9550                              Cafe                    55  

[9542 rows x 4 columns]

In [41]:
# Checking for missing values
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Cuisines                0
Average Cost for two    0
dtype: int64

In [42]:
df.shape

(9542, 4)

In [43]:
missing_values = df.isnull().sum()

In [44]:
print(missing_values)

Restaurant ID           0
Restaurant Name         0
Cuisines                0
Average Cost for two    0
dtype: int64


In [45]:


# Encode categorical variables
le_cuisines = LabelEncoder()
df['Cuisines'] = le_cuisines.fit_transform(df['Cuisines'])


In [48]:
df.columns

Index(['Restaurant ID', 'Restaurant Name', 'Cuisines', 'Average Cost for two'], dtype='object')

In [49]:
X = df.drop(['Cuisines', 'Restaurant ID', 'Restaurant Name'], axis=1)
y = df['Cuisines']

In [50]:

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)



In [51]:
# Step 3: Select a Classification Algorithm
# Initialize RandomForestClassifier
rf_classifier = RandomForestClassifier(random_state=42)

In [52]:
# Step 4: Train the Model
# Fit the model on the training data
rf_classifier.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [ ]:
# Step 5: Make Predictions
# Predict the values on the test set
y_pred = rf_classifier.predict(X_test)

In [53]:
# Calculate classification metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [54]:
# Display classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           1       0.00      0.00      0.00         1
           5       0.00      0.00      0.00         0
           6       0.02      0.33      0.03         3
          16       0.00      0.00      0.00         3
          18       0.00      0.00      0.00         1
          21       0.00      0.00      0.00         1
          25       0.00      0.00      0.00         1
          29       0.00      0.00      0.00         2
          35       0.00      0.00      0.00         2
          38       0.00      0.00      0.00         1
          41       0.00      0.00      0.00         1
          52       0.00      0.00      0.00         1
          54       0.00      0.00      0.00         4
          55       0.00      0.00      0.00         2
          58       0.00      0.00      0.00        16
          59       0.00      0.00      0.00         1
          63       0.00      0.00      0.00         1
    

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.

In [55]:
# Step 6: Analyze Performance
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)

Accuracy: 0.1288632792037716
Precision: 0.040329963624501065
Recall: 0.1288632792037716


In [57]:
from sklearn.linear_model import LogisticRegression

# Step 3 (Logistic Regression): Select a Classification Algorithm
# Initialize Logistic Regression
logistic_classifier = LogisticRegression(max_iter=1000, random_state=42)



In [ ]:

# Step 4 (Logistic Regression): Train the Model
# Fit the logistic regression model on the training data
logistic_classifier.fit(X_train, y_train)



In [62]:
# Step 5 (Logistic Regression): Evaluate the Model
# Predict on the testing data
y_pred_logistic = logistic_classifier.predict(X_test)



In [63]:
 #Calculate classification metrics for logistic regression
accuracy_logistic = accuracy_score(y_test, y_pred_logistic)
precision_logistic = precision_score(y_test, y_pred_logistic, average='weighted')
recall_logistic = recall_score(y_test, y_pred_logistic, average='weighted')



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [64]:
# Display classification report for logistic regression
print("Logistic Regression Classification Report:")
print(classification_report(y_test, y_pred_logistic))

# Step 6 (Logistic Regression): Analyze Performance
# You can further analyze the performance of logistic regression as needed.
print("Logistic Regression Accuracy:", accuracy_logistic)
print("Logistic Regression Precision:", precision_logistic)
print("Logistic Regression Recall:", recall_logistic)

# Compare the performance of Random Forest and Logistic Regression
print("\nComparison of Random Forest and Logistic Regression:")
print("Random Forest Accuracy:", accuracy)
print("Logistic Regression Accuracy:", accuracy_logistic)

Logistic Regression Classification Report:
              precision    recall  f1-score   support

           1       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         3
          16       0.00      0.00      0.00         3
          18       0.00      0.00      0.00         1
          21       0.00      0.00      0.00         1
          25       0.00      0.00      0.00         1
          29       0.00      0.00      0.00         2
          35       0.00      0.00      0.00         2
          38       0.00      0.00      0.00         1
          41       0.00      0.00      0.00         1
          52       0.00      0.00      0.00         1
          54       0.00      0.00      0.00         4
          55       0.00      0.00      0.00         2
          58       0.00      0.00      0.00        16
          59       0.00      0.00      0.00         1
          63       0.00      0.00      0.00         1
          64       0.00      0.00     

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [73]:
from sklearn.metrics import classification_report, confusion_matrix

# Compute the confusion matrix for Logistic Regression
cfm_logistic = confusion_matrix(y_test, y_pred_logistic)

In [74]:
# Compute the confusion matrix for Random Forest
cfm_random_forest = confusion_matrix(y_test, y_pred)

In [75]:
# Print the classification report and accuracy for Logistic Regression
print("Logistic Regression Classification Report:")
print(classification_report(y_test, y_pred_logistic))
accuracy_logistic = cfm_logistic.trace() / cfm_logistic.sum()

Logistic Regression Classification Report:
              precision    recall  f1-score   support

           1       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         3
          16       0.00      0.00      0.00         3
          18       0.00      0.00      0.00         1
          21       0.00      0.00      0.00         1
          25       0.00      0.00      0.00         1
          29       0.00      0.00      0.00         2
          35       0.00      0.00      0.00         2
          38       0.00      0.00      0.00         1
          41       0.00      0.00      0.00         1
          52       0.00      0.00      0.00         1
          54       0.00      0.00      0.00         4
          55       0.00      0.00      0.00         2
          58       0.00      0.00      0.00        16
          59       0.00      0.00      0.00         1
          63       0.00      0.00      0.00         1
          64       0.00      0.00     

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [76]:
print("Logistic Regression Accuracy:", accuracy_logistic)

Logistic Regression Accuracy: 0.08590885280251441


In [77]:
# Print the classification report and accuracy for Random Forest
print("\nRandom Forest Classification Report:")
print(classification_report(y_test, y_pred))
accuracy_random_forest = cfm_random_forest.trace() / cfm_random_forest.sum()



Random Forest Classification Report:
              precision    recall  f1-score   support

           1       0.00      0.00      0.00         1
           5       0.00      0.00      0.00         0
           6       0.02      0.33      0.03         3
          16       0.00      0.00      0.00         3
          18       0.00      0.00      0.00         1
          21       0.00      0.00      0.00         1
          25       0.00      0.00      0.00         1
          29       0.00      0.00      0.00         2
          35       0.00      0.00      0.00         2
          38       0.00      0.00      0.00         1
          41       0.00      0.00      0.00         1
          52       0.00      0.00      0.00         1
          54       0.00      0.00      0.00         4
          55       0.00      0.00      0.00         2
          58       0.00      0.00      0.00        16
          59       0.00      0.00      0.00         1
          63       0.00      0.00      0.00

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.

In [78]:
print("Random Forest Accuracy:", accuracy_random_forest)

Random Forest Accuracy: 0.1288632792037716


In [80]:
# Compare confusion matrices
print("\nComparison of Confusion Matrices:")
print("Logistic Regression Confusion Matrix:")
print(cfm_logistic)
print("\nRandom Forest Confusion Matrix:")
print(cfm_random_forest)


Comparison of Confusion Matrices:
Logistic Regression Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]

Random Forest Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 1 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 1 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [81]:
print("Logistic Regression Accuracy:", accuracy_logistic)
print("Random Forest Accuracy:", accuracy_random_forest)

Logistic Regression Accuracy: 0.08590885280251441
Random Forest Accuracy: 0.1288632792037716



Based on the accuracy scores obtained from the confusion matrices:

Logistic Regression Accuracy: 0.0859
Random Forest Accuracy: 0.1289
It's evident that the Random Forest model outperforms the Logistic Regression model in terms of accuracy. The Random Forest model achieved a higher accuracy of approximately 12.89%, compared to the Logistic Regression model's accuracy of approximately 8.59%.

This suggests that the Random Forest algorithm may be better suited for this classification task compared to Logistic Regression.

In summary, based on the accuracy comparison, the Random Forest model shows better performance than the Logistic Regression model for classifying restaurants based on their cuisines.